### PR TITLE
"dry" up hydrated resources by default

### PR DESCRIPTION
**Pull request by jakswa**
- taking another path to solve https://app.shortcut.com/grayscale/story/8428/apply-memory-savings-changes-to-greenhouse-client-gem

This ended up much quicker than the alternative "just keep a single page around" I implemented elsewhere. In the gem we have two `include Enumerable` classes meshed together here that are both hydrating arrays.

This will break any enum usage that expects actual records in hydrated arrays. It pretty much only keeps `.count` (with _no arguments_) accurate. At least I would be very surprised if I couldn't run normal methods in a `.count { |record| record.method }` situation.

### CODE REVIEW

File: spec/greenhouse_io/api/candidate_collection_spec.rb [hunk 1]
```diff
@@ -37,4 +37,17 @@
...
+    expect(all[0]).to be_a(GreenhouseIo::Candidate)
+    expect(candidates.count {|i| i == :dried }).to eq candidates.count
+  end
+
+  it 'blanks out pages when .each is called' do
+    all = candidates.each_page.to_a
+    expect(all[0][0]).to be_a(GreenhouseIo::Candidate)
+    all[0].each {}
+    expect(all[0][0]).to eq :dried
+  end
```
Comment (jakswa) [inline spec/greenhouse_io/api/candidate_collection_spec.rb:52]: I tried to demonstrate/document the surprising behavior in a couple specs.

File: lib/greenhouse_io/api/resource_collection.rb [hunk 2]
```diff
@@ -18,12 +18,14 @@ class ResourceCollection
...
+    def initialize(client:, query_params: {}, resource_class:, dry: true)
       self.client             = client
       self.resource_class     = resource_class # e.g. GreenhouseIo::Application
       self.lazy_paginators    = [LazyPaginator.new(resource_collection: self, query_params: query_params)]
       self.hydrated_resources = []
       self.hydrated_pages     = []
+      @dry = dry
     end
 
     def each_page
```
Comment (jakswa) [inline lib/greenhouse_io/api/resource_collection.rb:31]: do we use `each_page` anywhere in our code? Depending on the usage, we might be keeping records around. Currently the cleanup on pages is only happening when the page is iterated through (`Page#each`).

Comment (etipton) [reply lib/greenhouse_io/api/resource_collection.rb:31]: Looks like we do, introduced in: https://github.com/Grayscale-Labs/cATs/pull/289

File: lib/greenhouse_io/api/resource_collection.rb [hunk 3]
```diff
@@ -18,12 +18,14 @@ class ResourceCollection
 
     attr_accessor :client, :resource_class
 
-    def initialize(client:, query_params: {}, resource_class:)
+    # @param dry [Boolean] When true, we try to keep only `:dried` around in the hydration arrays
```
Comment (etipton) [inline lib/greenhouse_io/api/resource_collection.rb:21]: Love the idea of using `:dried` instead of `nil` (even better would be a "dummy class", e.g. `LazyPaginator::DryResource`)

WDYT of being more explicit w/ this param name, e.g. `dehydrate_after_iteration` or something?

File: lib/greenhouse_io/api/resource_collection.rb [hunk 4]
```diff
@@ -18,12 +18,14 @@ class ResourceCollection
...
+    def initialize(client:, query_params: {}, resource_class:, dry: true)
       self.client             = client
       self.resource_class     = resource_class # e.g. GreenhouseIo::Application
       self.lazy_paginators    = [LazyPaginator.new(resource_collection: self, query_params: query_params)]
       self.hydrated_resources = []
       self.hydrated_pages     = []
+      @dry = dry
```
Comment (etipton) [inline lib/greenhouse_io/api/resource_collection.rb:28]: Any opposition to using attr_accessor like w/ the other params? FWIW, I used to think it felt too "formal" to do that, but then I read a blog post similar to [this one](https://ivoanjo.me/blog/2017/09/20/why-i-always-use-attr_reader-to-access-instance-variables/) which convinced me to pretty much avoid ivars (aside from memoization) -- overall though, DWYTIB, this is just a "nit" comment

**Review by etipton — CHANGES_REQUESTED** [review]
couple thoughts, most significant is the response RE: `each_page`

**Comment by jakswa** [comment]
The latest commit renames the variable and adopts the `attr_accessor` in classes using that. I spent some time trying clean up pages, but that code got hard to follow and I stared at our `each_page` usage and verified we are calling `Page#each`.

**Review by kphan32 — APPROVED** [review]
lgtm :+1:
